File: strategy/java/CustomerBill.java
import java.util.Vector;
import java.util.List;

interface BillingStrategy {
    // Use a price in cents to avoid floating point round-off error
    int getActPrice(int rawPrice);
  
    // Normal billing strategy (unchanged price)
    static BillingStrategy normalStrategy() {
        return rawPrice -> rawPrice;
    }
  
    // Strategy for Happy hour (50% discount)
    static BillingStrategy happyHourStrategy() {
        return rawPrice -> rawPrice / 2;
    }
}

final class CustomerBill {

    private final List<Integer> drinks = new Vector<>();
    private BillingStrategy strategy;

    public CustomerBill(BillingStrategy strategy) {
        this.strategy = strategy;
    }

    public void add(int price, int quantity) {
        this.drinks.add(this.strategy.getActPrice(price*quantity));
    }

    // Payment of bill
    public void print() {
        int sum = this.drinks.stream().mapToInt(v -> v).sum();
        System.out.println("Total due: " + sum);
        this.drinks.clear();
    }

    // Set Strategy
    public void setStrategy(BillingStrategy strategy) {
        this.strategy = strategy;
    }

    public static void main(String... arguments) {
        // Prepare strategies
        BillingStrategy normalStrategy = BillingStrategy.normalStrategy();
        BillingStrategy happyHourStrategy = BillingStrategy.happyHourStrategy();
        BillingStrategy freeStrategy = rawPrice -> 0;

        CustomerBill firstCustomer = new CustomerBill(normalStrategy);

        // Normal billing
        firstCustomer.add(100, 1);

        // Start Happy Hour
        firstCustomer.setStrategy(happyHourStrategy);
        firstCustomer.add(100, 2);

        // New Customer
        CustomerBill secondCustomer = new CustomerBill(happyHourStrategy);
        secondCustomer.add(80, 1);
        // The Customer pays
        firstCustomer.print();

        // Happy Hour
        secondCustomer.setStrategy(normalStrategy);
        secondCustomer.add(130, 2);
        secondCustomer.add(250, 1);
        secondCustomer.print();

        // Free Strategy
        secondCustomer.setStrategy(freeStrategy);
        secondCustomer.add(130, 2);
        secondCustomer.add(250, 1);
        secondCustomer.add(300, 5);
        secondCustomer.add(250, 1);
        secondCustomer.print();
    }
}

File: strategy/kotlin/customer_bill.kt
class CustomerBill(var strategy: (Int) -> Int) {

    private val drinks: MutableList<Int> = ArrayList()

    fun add(price: Int, quantity: Int): Unit {
        this.drinks.add(this.strategy(price * quantity))
    }

    fun print(): Unit {
        val sum: Int = this.drinks.sumBy { it } ;
        println("Total due: " + sum);
        this.drinks.clear();
    }
}

fun main(): Unit {
    // Prepare strategies
    val normalStrategy: (Int) -> Int = { it }
    val happyHourStrategy: (Int) -> Int = { it / 2 }
    val freeStrategy: (Int) -> Int = { 0 }

    val firstCustomer: CustomerBill = CustomerBill(normalStrategy);

    // Normal billing
    firstCustomer.add(100, 1);

    // Start Happy Hour
    firstCustomer.strategy = happyHourStrategy;
    firstCustomer.add(100, 2);

    // New Customer
    val secondCustomer: CustomerBill = CustomerBill(happyHourStrategy);
    secondCustomer.add(80, 1);
    // The Customer pays
    firstCustomer.print();

    // Happy Hour
    secondCustomer.strategy = normalStrategy;
    secondCustomer.add(130, 2);
    secondCustomer.add(250, 1);
    secondCustomer.print();

    // Free Strategy
    secondCustomer.strategy = freeStrategy;
    secondCustomer.add(130, 2);
    secondCustomer.add(250, 1);
    secondCustomer.add(300, 5);
    secondCustomer.add(250, 1);
    secondCustomer.print();
}